soroban-env-host/src/crypto/bls12_381.rs

```diff
@@ -0,0 +1,665 @@
+                let mut expected_bytes = [0; EXPECTED_SIZE];
+                expected_bytes[0] = flags;
+                if bytes != expected_bytes {
+                    Err(self.err(ScErrorType::Crypto, ScErrorCode::InvalidInput, format!("bls12-381 {msg} deserialize: infinity flag (bit 1) is set while remaining bits are not all zero").as_str(), &[]))
+                } else {
+                    Ok(())
+                }
+            },
+            0b0000_0000 => Ok(()), // infinite bit is unset
+            _ => Err(self.err(ScErrorType::Crypto, ScErrorCode::InvalidInput, format!("bls12-381 {msg} deserialize: either compression flag (bit 0) or the sort flag (bit 2) is set, while the input should be encoded uncompressed").as_str(), &[]))
+        }
+    }
+
+    pub(crate) fn metered_check_point<P: SWCurveConfig>(
+        &self,
+        pt: Affine<P>,
+        ty: ContractCostType,
+    ) -> Result<Affine<P>, HostError> {
+        self.charge_budget(ty, None)?;
+        // performs following checks 1. point belongs to the curve 2. point
+        // belongs to the correct subgroup
+        pt.check().map_err(|_| {
+            self.err(
+                ScErrorType::Crypto,
+                ScErrorCode::InvalidInput,
+                "bls12-381 G1 affine deserialize: invalid point",
+                &[],
+            )
+        })?;
+        Ok(pt)
+    }
+
+    pub(crate) fn g1_affine_deserialize_from_bytesobj(
+        &self,
+        bo: BytesObject,
+    ) -> Result<G1Affine, HostError> {
+        let msg: &str = "G1";
+        let pt: G1Affine = self.visit_obj(bo, |bytes: &ScBytes| {
+            self.validate_point_encoding::<G1_SERIALIZED_SIZE>(&bytes, msg)?;
+            // `CanonicalDeserialize` of `Affine<P>` calls into
+            // `P::deserialize_with_mode`, where `P` is `arc_bls12_381::g1::Config`, the
+            // core logic is in `arc_bls12_381::curves::util::read_g1_uncompressed`.
+            //
+            // The `arc_bls12_381` lib already expects the input to be serialized in
+            // big-endian order (aligning with the common standard and contrary
+            // to ark::serialize's convention),
+            //
+            // i.e. `input = be_bytes(X) || be_bytes(Y)` and the
+            // most-significant three bits of X are flags:
+            //
+            // `bits(X) = [compression_flag, infinity_flag, sort_flag, bit_3, .. bit_383]`
+            //
+            // internally when deserializing `Fp`, the flag bits are masked off
+            // to get `X: Fp`. The Y however, does not have the top bits masked off
+            // so it is possible for Y to exceed 381 bits. I've checked all over and
+            // didn't find that being an invalid condition, so we will leave them as is.
+            self.deserialize_uncompessed_no_validate(&bytes, 2, msg)
+        })?;
+        self.metered_check_point::<ark_bls12_381::g1::Config>(
```

Comment:
What is the consequence of not doing this check for each of the bls operations? For example, what happens if we pass an invalid point to the MSM host function?

Reply:
For multiplication, the underlying algorithm uses the endomorphism optimization, which basically performs mapping of a point from one subgroup to another subgroup of the same size. So it is necessary for the point to belong to the subgroup. If not the underlying algorithm will likely fail.